---
title: "R package Tutorial"
author: "Marta Cousido Rocha"
date: "7/1/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Instructions for Creating an R Package

The following is a step-by-step guide to creating your own R package using Rstudio.


### Creating the files and the new project

- Start by opening a new .R file. Make sure your default directory is clear by typing rm(list
= ls()). 
- Write the code for your functions in this .R file. You can create one file with all of your functions or create separate files for each function. Save these files.

In this tutorial we start creating an R file "square.R" containing the following R function.

```
square<-function(x){
 y<-x^2
  return(y)
}
```


- Install the ‘devtools’ package 

```
install.packages("devtools")
```

R package development has become substantially easier in recent years with the introduction of this package. As the package name suggests, it includes a variety of functions that facilitate software development in R.


- Open a new project in RStudio. Go to the ‘File’ menu and click on ‘New Project.’ Then
select ‘New Directory,’ and ‘R Package’ to create a new R package.


- Type the name of your package, then upload the .R file that you created before. Click ‘Create project.’



### Structure of the  project

On the lower right hand side of your screen, you should see a file directory.

- A file named DESCRIPTION containing explanations of the package, author, and license conditions
in a structured text format.


- A file named NAMESPACES, this file 
ensure that other packages will not interfere with your code and the package works regardless of the environment in which it’s run.
More precisely, this file allow the package writer to hide functions and data that are meant only for internal use,
and prevent functions from breaking when a user (or other package writers) picks a name that clashes with one in the package.

- The ‘R’ folder
contains the code for your functions. 

- The ‘man’ folder will contain the help files for each
function in your package. Depending on your version of RStudio, the help files may have been
generated automatically as .Rd or “R documentation” files when you created your package.




### Roxygen2 package

Roxygen2 package is useful for automating the process fo generating .Rd (and even
NAMESPACE) files.
With Roxygen2, you write specially-structured comments preceding each function definition. These are processed to produce the .Rd files that R needs, and it also creates the NAMESPACE file for you. Then, install this package and load it. 

```
install.packages("roxygen2")
```

Go to 'Tools', 'Version Control' and 'Project Setup'. Go to 'Build Tools', in the new window select 'Package' in 'Project build tools', then click in 'Use devtools package functions if available', and finally click in 'Generate documentation with Roxygen2' and select all the items.


The Roxygen2 comments are just R comments (preceded by #), but you need to use #' to distinguish the Roxygen2 comments from any normal R comments. 

Open the "square.R" file and includes the following Roxygen2 comments.


```javasript
#' @title A function to square a number
#'
#' @description
#' Performs the square of a number.
#'
#' @usage
#' square(x)
#'
#' @param x a number to square (see details).
#' 
#' @details
#' The function computes the square of a number, x must be numerical. Check that x is not a character or factor object.
#'
#' @return y which is the square of x.
#'
#'
#' @author
#' \itemize{
#' \item{Marta Cousido-Rocha}
#' \item{Santiago Cerviño}
#' }
#'
#' @references
#' \itemize{
#' \item{Here you can write a reference related to the function.}
#' }
#'
#' @examples
#'
#' x<-2; square(x)
#' @export
```

### DESCRIPTION File

The Package, Version, License, Description, Title, Author, and Maintainer
fields are mandatory, the remaining fields (Date, Suggests, ...) are optional.

- The Package and Version fields give the name and the version of the package, respectively.

- The name should consist of letters, numbers, and the dot character and start with a letter. 

- The
version is a sequence of at least two (and usually three) non-negative integers separated by single
dots or dashes.

- The Title should be no more than 65 characters, because it will be used in various package
listings with one line per package. 

- The Author field can contain any number of authors in free text
format, the Maintainer field should contain only one name plus a valid email address (similar
to the “corresponding author” of a paper). 

- The Description field can be of arbitrary length.

- The License can be free text, if you submit the package to CRAN or Bioconductor and
use a standard license, we strongly prefer that you use a standardized abbreviation like GPL-2
which stands for “GNU General Public License Version 2”. 

A basis DESCRIPTION file is the following.
```
Package: square
Type: Package
Title: Square a number
Version: 0.1.0
Author: Marta Cousido [aut, cre]
Maintainer: Marta Cousido Rocha <marta.cousido@ieo.es>
Description: The package allow to compute the square of a number.
License: GPL-2
Encoding: UTF-8
LazyData: true
```

To convert our package to use Roxygen2, we will want to first remove that minimal NAMESPACE file we had created. Then load the devtools package with library(devtools) and type

```
document()
```

You will then see that your package directory contains a NAMESPACE file and a man/ subdirectory with an .Rd file for each of the documented functions.

That’s it! Your package is now documented.

### Compiling our package

Go to 'Build' and choose 'Check' for running all sorts of checks on the contents of an R package, and gives warnings and error messages when it finds things that aren’t right. It also will run the examples in the .Rd files for each of your functions. If no errors, warnings or notes appears you are ready to compile your package. Go to ‘Build’ on the top toolbar and select ‘Install and Restart’ (note you can also use the keyboard shortcut Ctrl+Shift+B). If this works, your
package will automatically load and you will see library(square) at the bottom of
your console. Test your functions to make sure they work. Firstly, we can see the help page of our function as follows.

```
?square
```

Then, run the example to check that the function really works.


If you want to distribute your package to an R user before to upload it to CRAN or Github, you will need to make a source package. Go to 'Build' and then 'Build Source Package' to create a .tar.gz file.


###  Useful information

#### *Keyboard shortcuts*

- CONTROL + SHIFT + d: Update the documentation.
- CONTROL + SHIFT + e: Executate the basic checkings of the package.
- CONTROL + SHIFT + b: Try to install the package, and if it works you can check the different functions.


#### *Basic Tricks*

- Write <- instead of = for assignments.
- Put TRUE and FALSE instead of T/F.
- Reference different functions such as rnorm, var, sd ,... using stats ::
rnorm (). The same for the graphic elements plot, lines, ... using graphics ::
lines(). This is a very common mistake. In the libraries that you import, it is not
necessary. 

Open the square.R file, and write 

```
#' @importFrom stats var
```
and change the square function as follows

```
square<-function(x){
  y<-x^2
  z<-var(c(x,x^2))
  return(c(y,z))
}
```

- If the output of the function is the result of a test, you can assign it to the final list the class htest: class (object) <- "htest" and customize the output
so that it has the form of a test. 

- If you do not want to export a function just skip the description.

#### *Data*

- Open the data in R.
- Save them as .rda with save (\ data.rda).
- Save the files in the /data folder (if you don't have create it) and create R files with the documentation
on of each data set.

Create the following R file in the R folder, and include in the folder data the .rda file containing Hedenfalk data.

See the use of dontshow and donttest in the following file.
```
#' Hedenfalk data
#'
#' These data come from the breast cancer gene expression study of Hedenfalk et al. (2001). The data set
#' consists on 3226 gene expression levels measured on 7 patients with breast tumors having BRCA1
#' mutations (columns 1-7) and on 8 patients with breast tumors having BRCA2 mutations (columns 8-15).
#' The rows correspond to the genes and the columns refer to the patients.
#' @docType data
#' @usage data(Hedenfalk)
#'
#' @format A matrix with 3226 rows corresponding to the measured genes and 15 columns corresponding to the
#' patients. The first 7 columns contain the measures for the patients with BRCA1 mutations and the
#' patients with BRCA2 mutations are located in the last 8 columns.
#'
#'
#' @source \url{http://genomine.org/qvalue/results.txt}
#'
#' @references
#'
#' Hedenfalk, I., Duggan, D., Chen, Y., Radmacher, M, Bitter, M., Simon, R., Meltzer, P., Guterson, B., Esteller, M., Kallioniemi, O., Wilfond, B., Borg, A., Trent, J., Raffeld, M., Yakhini, Z., BenDor, A., Dougherty, E., Kononen, J., Buberdorf, L., Fehrle, W., Pittaluga, S., Gruvberger, G., Loman, N., Johannsson, O., Olsson, H., and Sauter, G.(2001), Gene-Expression Profiles in Hereditary Breast Cancer. New England Journal of Medicine 344 (8), 539-548.
#'
#' @examples
#' \dontshow{
#' data(Hedenfalk)
#' X <- Hedenfalk
#' }
#'
#' \donttest{
#' data(Hedenfalk)
#' X <- Hedenfalk}
```


#### *CRAN*

If you want to upload the package to CRAN you must do the check of the package as CRAN (which is not exactly the same as with
the keyboard shortcut). You have to go to the RStudio terminal, go to the folder
where is the package (with the cd command) and write:

R CMD check --as-cran square.tar.gz

This check will always leave 1 Note that is unrelevant (when uploading
to CRAN also in).

Reminder: copy and paste each error, warning or note that you can get in
Google. In stackover is usually the answer.